src/vs/workbench/contrib/notebook/browser/contrib/notebookVariables/notebookInlineVariables.ts

```diff
@@ -0,0 +1,356 @@
+import { ServicesAccessor } from '../../../../../../platform/instantiation/common/instantiation.js';
+import { IDebugService } from '../../../../debug/common/debug.js';
+import { Expression } from '../../../../debug/common/debugModel.js';
+import { NotebookSetting } from '../../../common/notebookCommon.js';
+import { ICellExecutionStateChangedEvent, INotebookExecutionStateService } from '../../../common/notebookExecutionStateService.js';
+import { INotebookKernelMatchResult, INotebookKernelService, VariablesResult } from '../../../common/notebookKernelService.js';
+import { INotebookActionContext, NotebookAction } from '../../controller/coreActions.js';
+import { ICellViewModel, INotebookEditor, INotebookEditorContribution } from '../../notebookBrowser.js';
+import { registerNotebookContribution } from '../../notebookEditorExtensions.js';
+
+// value from debug, may need to keep an eye on and shorter to account for cells having a narrower viewport width
+const MAX_INLINE_DECORATOR_LENGTH = 150; // Max string length of each inline decorator. If exceeded ... is added
+
+const variableRegex = new RegExp(
+	'(?:[a-zA-Z_][a-zA-Z0-9_]*\\.)*' + //match any number of variable names separated by '.'
+	'[a-zA-Z_][a-zA-Z0-9_]*', //math variable name
+	'g',
+);
+
+class InlineSegment {
+	constructor(public column: number, public text: string) {
+	}
+}
+
+export class NotebookInlineVariablesController extends Disposable implements INotebookEditorContribution {
+
+	static readonly id: string = 'notebook.inlineVariablesController';
+
+	private cellDecorationIds = new Map<ICellViewModel, string[]>();
+	private cellContentListeners = new ResourceMap<IDisposable>();
+
+	constructor(
+		private readonly notebookEditor: INotebookEditor,
+		@INotebookKernelService private readonly notebookKernelService: INotebookKernelService,
+		@INotebookExecutionStateService private readonly notebookExecutionStateService: INotebookExecutionStateService,
+		@ILanguageFeaturesService private readonly languageFeaturesService: ILanguageFeaturesService,
+		@IConfigurationService private readonly configurationService: IConfigurationService,
+		@IDebugService private readonly debugService: IDebugService,
+	) {
+		super();
+
+		this._register(this.notebookExecutionStateService.onDidChangeExecution(async e => {
+			if (!this.configurationService.getValue<boolean>(NotebookSetting.notebookInlineVariables)) {
+				return;
+			}
+
+			if (this.isCellExecutionStateChangedEvent(e)) {
+				await this.updateInlineVariables(e);
+			}
+		}));
+	}
+
+	private async updateInlineVariables(event: ICellExecutionStateChangedEvent): Promise<void> {
+		if (this.debugService.state !== 0) { // debug session is a state other than inactive, so clear and defer to whatever it's handling
+			this._clearNotebookInlineDecorations();
+			return;
+		}
+
+		if (this.notebookEditor.textModel?.uri !== event.notebook) {
```

Comment:
`this.notebookEditor.textModel?.uri && isEqual(this.notebookEditor.textModel.uri, event.notebook)`